### ISSUE TITLE
connect swallows errors in mapStateToProps

### ISSUE THREAD

**Issue by grumd** [reduxjs/react-redux#1942]
### What version of React, ReactDOM/React Native, Redux, and React Redux are you using?

- React: 18.2
- ReactDOM/React Native: react-dom 18.2
- Redux: 4.2.0
- React Redux: 8.0.2

### What is the current behavior?

https://codesandbox.io/s/gifted-architecture-wr4f86?file=/src/App.js

1. Have a code that can cause an error in mapStateToProps, e.g.
```js
const mapStateToProps = (state) => {
  return {
    foo: state.foo.length, // when foo is undefined, causes an error
  };
};
```
2. Dispatch an action that changes the state so that it causes the error above
```js
dispatch(updateFoo(undefined));
```
3. Rerender doesn't happen, no errors emitted in console, componentDidCatch catches nothing. The error is swallowed.

### What is the expected behavior?

This issue makes debugging code that's using react-redux really hard.
The expected behaviour is that the error is visible in console.

### Which browser and OS are affected by this issue?

Windows, latest Chrome here. Probably affects all browsers.

### Did this work in previous versions of React Redux?

- [ ] Yes

**Comment by grumd** [comment]
Just tested - it worked fine in 7.0.3

Here's a 7.0.3 fork https://codesandbox.io/s/charming-firefly-coekps

**Comment by markerikson** [comment]
This is likely due to the switch to `useSyncExternalStore`, which uses the same approach that `useSelector` does in trapping errors.

I'd have to investigate at some point, but I'm not sure how much we'll be able to do here, and any further changes to `connect` are _really_ low priority.

Honestly the main advice here would be to _not_ use `connect`, and switch to `useSelector` instead.

**Comment by grumd** [comment]
React 18 still supports class components, and you'll need react-redux 8 for React 18.
It's not the biggest blocking bug of course, but still would be nice if react-redux wasn't preventing people from upgrading to React 18 when their codebase has legacy class components.

I understand your point though. Would be great if you could take a look someday! Thank you for your work.

**Comment by markerikson** [comment]
Sure. To be clear, I wouldn't see this as "blocking an upgrade" to React-Redux v8 + React 18.  We actually are on both in https://github.com/replayio/devtools right now, and we _do_ have a bunch of `connect`ed class components in our codebase still. So, I know what that's like :)  It's "just" that if there's an error in `mapState` it's not getting logged. Totally get that that's a pain, and I'm not trying to diminish that aspect - just noting that in terms of "does the code run and do the libs work", the answer is definitely "yes".

**Comment by grumd** [comment]
Yeah I completely agree, don't get me wrong. Just leaving a little bit of hope that maybe you'll be bored one day and will fix it anyway :)

**Comment by itsjohncs** [comment]
Just ran into this in `7.2.5` (`react@17.0.2`, `react-redux@7.2.5`, and `redux@4.1.1`). So the story of which versions are affected seems to be more muddled.

**Comment by rdonnelly** [comment]
I think we're running into something similar in our application. When we hit an error in a selector, the exception seems to get gobbled up here: 

https://github.com/reduxjs/react-redux/blob/8cf538c05b21433eba2106a6609c6ccbfce471de/src/components/connect.tsx#L131-L141

We were previously relying on that thrown error to bubble up and get caught by an error handler, but now our app won't re-render and won't show an error. I won't claim that this is the _best_ pattern (the application code is pretty old at this point), but is there any way to let that error pass through? Or if we did switch to `useSelector` syntax rather than `connect`, would that alleviate our issues?

**Comment by markerikson** [comment]
@rdonnelly : afraid I don't have the brain bandwidth to answer this one further atm. The logic in `connect` is frankly very fragile, doubly so in v8, and we're not going to make further changes to it.  I'd suggest trying out `useSelector` and see what happens.

**Comment by alexanderchr** [comment]
Just spent hours trying to figure out why my app was silently crashing and this was the reason. In my opinion this is a quite significant regression, at least enough for me to have to downgrade to 7.1.x for now. Clearly `useSelector` is the way to go but I cannot justify removing all `connect`s right now.

I really wish you'd reconsider the decision to not make further changes to `connect`, but understand it might not be an option. At the very minimum it should be documented because it `is` a breaking change from 7.1.x even if it does not break any happy paths.

**Comment by alexanderchr** [comment]
Just a thought, would it be possible to log the error here if `NODE_ENV !== production`?

> https://github.com/reduxjs/react-redux/blob/8cf538c05b21433eba2106a6609c6ccbfce471de/src/components/connect.tsx#L131-L141

Would still be breaking for anyone whose logic depends upon errors in selectors bubbling up but hopefully that's not very many. Not perfect but a minimal change that would at least make it easier to find the cause of silent crashes caused by this during development.

**Comment by markerikson** [comment]
@alexanderchr I'm open to PRs to improve this, but right now all the time and mental bandwidth I can scrape together for Redux work is focused on RTK 2.0 and the related major versions, so I don't have the energy to think about this one myself.

**Comment by alexanderchr** [comment]
@markerikson I appreciate that, at first I took it as if you were not even considering PRs. Fully understand that this is not your top priority.

I think that unless you have a full formal specification of a library there will always be lots of common sense behavior that is nowhere documented but cannot be changed without causing a breaking change. I think that not swallowing errors in functions passed to the library is usually such behavior, but obviously understand that it's impossible to keep track of it all (and also a question of personal judgement).

**Comment by hipstersmoothie** [comment]
Any updates on this? This is blocking our upgrade path

**Comment by markerikson** [comment]
@hipstersmoothie : no updates . Per above, all my efforts in the last year+ have gone to upgrading the Redux packages (RTK 2.0, etc), and reworking the "Redux Essentials" tutorial.

I'm not ruling out future changes to `connect` completely, but working on it is _not_ a priority any time in the near future.

If you or someone else would like to submit a PR, I'm open to looking at it.

What specifically about the `connect` error handling behavior is causing you problems at this time?  

(Also, is there a specific reason why you _can't_ migrate to `useSelector` in your codebase?)

**Comment by srubin** [comment]
(I work with @hipstersmoothie so I'll answer!)

We haven't actually been able to confirm that this bug is happening to us in 8.1.3. We're seeing errors getting bubbled up to our react error boundaries as expected, just like in react-redux 7.x. The codesandbox links in the original post seem to be dead now, so I can't test the original repro case. I'm wondering if there was some specific context around when it failed that we're not capturing in our synthetic tests. So, hopefully this is a false alarm on our end!

> (Also, is there a specific reason why you can't migrate to useSelector in your codebase?)

Mostly the risks of large-scale refactoring jobs in complex codebases, i.e., ensuring that nothing breaks along the way, and that performance is not regressing. We have some fun custom wrappers for `connect`, so we'd need to add hook versions of those changes, as well. So, it's not that we _can't_ do it, it's just that it'd be a nontrivial project that needs to be weighed against all of our other priorities.

**Comment by markerikson** [comment]
Yeah, I definitely understand the effort to do major migrations :)

Out of curiosity, what sort of "custom `connect` wrappers" do you have?  What do those look like?

**Comment by srubin** [comment]
> Out of curiosity, what sort of "custom `connect` wrappers" do you have? What do those look like?

One is related to error-tracking (we want errors thrown in mapStateToProps to be tracked to Sentry with specific additional metadata).

The more interesting one modifies the behavior of `mapStateToProps` to preserve reference equality for values that are deeply equal to their previous results. While this can be addressed with selectors/cached selectors/the mstp factory pattern, doing this _everywhere_ that connect is used helps us avoid an entire class of unnecessary re-renders and simplifies code. Our data indicates that perf is as good or better than without this wrapper.

**Comment by alexanderchr** [comment]
Ok this bit me again yesterday so I took some time to look into it. Happy to share minimal repro if anyone is interested.

The problem is that the error is caught here:

https://github.com/reduxjs/react-redux/blob/7e2fdd4ee2021e4282e12ba9fc722f09124e30cd/src/components/connect.tsx#L133-L136

This was also the case in 7.2.9, however at that version, an error in a selector caused the component to dismount which in turn caused the error to be rethrown a few lines down:

https://github.com/reduxjs/react-redux/blob/7e2fdd4ee2021e4282e12ba9fc722f09124e30cd/src/components/connect.tsx#L170-L186

From 8.0.0, the component is no longer dismounted when there is an error in a selector, only rerendered in its pre-error state. Thus the error is silently swallowed.

A very simple fix would be to log the error message when it is caught in non-production environments. Would a PR doing this be considered?

**Comment by markerikson** [comment]
Yeah, seems like `useSyncExternalStore` handles the errors differently while rendering than we did with our own logic. I'm a bit surprised - I thought `uSES` pretty directly copied what we did.

I'd be curious to know what `uSES` _does_ with an error thrown while rendering, exactly, or if we're doing something different now that means it _doesn't_ get thrown.

But yeah, logging this in dev seems like a passable fallback.

**Comment by alexanderchr** [comment]
I'm also curious what is going on there. I'll do some more digging and share if I find anything.

**Comment by webdevelopersrinu** [comment]
Hi @markerikson @alexanderchr — happy to pick this up.

Planning the minimal dev-only logging approach you approved in your Dec 12, 2024 comment:
- Add a `console.error` in the catch block at `connect.tsx:133-136` guarded by `process.env.NODE_ENV !== 'production'`
- Add a test in `connect.spec.tsx` covering the case
- No production behavior change

If that direction still works for you, I'll open the PR.
